File: password-hash/password-hash.py
def authEvents(events):
    p = 131
    M = 1000000007

    password_hash = 0

    def hash_string(s):
        hash_val = 0
        for c in s:
            hash_val = (hash_val * p + ord(c)) % M
        return hash_val

    def authorize(x, current_hash):
        if x == current_hash:
            return 1
        for i in range(128):  # ASCII range
            new_hash = (current_hash * p + i) % M
            if new_hash == x:
                return 1
        return 0

    result = []
    for command, param in events:
        if command == "setPassword":
            password_hash = hash_string(param)
        elif command == "authorize":
            x = int(param)
            result.append(authorize(x, password_hash))

    return result

# Example usage:
events = [
    ["setPassword", "cAr1"],
    ["authorize", "223691457"],
    ["authorize", "303580761"]
]
print(authEvents(events))  # Expected output: [1, 1]

File: password-hash/password-hash.js
function authEvents(events) {
    const p = 131
    const M = Math.pow(10, 9) + 7

    const hashString = (s) => {
        let hash = 0
        for (let i = 0; i < s.length; i++) {
            hash = (hash * p + s.charCodeAt(i)) % M
        }
        return hash
    }

    const authorize = (x, currentHash) => {
        if (x === currentHash) return 1

        for (let i = 0; i < 128; i++) {
            let newHash = (currentHash * p + 1) % M
            if (newHash === x) return 1
        }
        return 0
    }

    const result = []
    for (let [command, param] of events) {
        if (command === "setPassword") {
            passwordHash = hashString(param)
        } else if (command === "authorize") {
            const x = parseInt(param)
            result.push(authorize(x, passwordHash))
        }
    }

    return result
}